File: java/src/main/java/cn/wjybxx/dson/text/CharBuffer.java
/*
 * Copyright 2023 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dson.text;

import cn.wjybxx.dson.internal.BinaryUtils;

import javax.annotation.Nonnull;
import java.nio.BufferOverflowException;
import java.util.Arrays;
import java.util.Objects;

/**
 * @author wjybxx
 * date - 2023/7/8
 */
class CharBuffer implements CharSequence {

    public char[] buffer;
    public int ridx;
    public int widx;

    CharBuffer(char[] buffer) {
        this.buffer = Objects.requireNonNull(buffer);
    }

    public int capacity() {
        return buffer.length;
    }

    public boolean isReadable() {
        return ridx < widx;
    }

    public boolean isWritable() {
        return widx < buffer.length;
    }

    public int writableChars() {
        return buffer.length - widx;
    }

    public int readableChars() {
        return Math.max(0, widx - ridx);
    }

    public char read() {
        if (ridx == widx) throw new BufferOverflowException();
        return buffer[ridx++];
    }

    public void unread() {
        if (ridx == 0) throw new BufferOverflowException();
        ridx--;
    }

    public void write(char c) {
        if (widx == buffer.length) {
            throw new BufferOverflowException();
        }
        buffer[widx++] = c;
    }

    public void write(char[] chars) {
        if (chars.length == 0) {
            return;
        }
        if (widx + chars.length > buffer.length) {
            throw new BufferOverflowException();
        }
        System.arraycopy(chars, 0, buffer, widx, chars.length);
        widx += chars.length;
    }

    public void write(char[] chars, int offset, int len) {
        if (len == 0) {
            return;
        }
        BinaryUtils.checkBuffer(chars.length, offset, len);
        if (widx + len > buffer.length) {
            throw new BufferOverflowException();
        }
        System.arraycopy(chars, offset, buffer, widx, len);
        widx += len;
    }

    /**
     * 将给定buffer中的可读字符写入到当前buffer中
     * 给定的buffer的读索引将更新，当前buffer的写索引将更新
     *
     * @return 写入的字符数；返回0时可能是因为当前buffer已满，或给定的buffer无可读字符
     */
    public int write(CharBuffer charBuffer) {
        int n = Math.min(writableChars(), charBuffer.readableChars());
        if (n == 0) {
            return 0;
        }
        write(charBuffer.buffer, charBuffer.ridx, n);
        charBuffer.addRidx(n);
        return n;
    }

    //

    public void addRidx(int count) {
        setRidx(ridx + count);
    }

    public void addWidx(int count) {
        setWidx(widx + count);
    }

    public void setRidx(int ridx) {
        if (ridx < 0 || ridx > widx) {
            throw new IllegalArgumentException("ridx overflow");
        }
        this.ridx = ridx;
    }

    public void setWidx(int widx) {
        if (widx < ridx || widx > buffer.length) {
            throw new IllegalArgumentException("widx overflow");
        }
        this.widx = widx;
    }

    public void setIndexes(int ridx, int widx) {
        if (ridx < 0 || ridx > widx) {
            throw new IllegalArgumentException("ridx overflow");
        }
        if (widx > buffer.length) {
            throw new IllegalArgumentException("widx overflow");
        }
        this.ridx = ridx;
        this.widx = widx;
    }

    //

    public void shift(int shiftCount) {
        if (shiftCount <= 0) {
            return;
        }
        if (shiftCount >= buffer.length) {
//                Arrays.fill(buffer, (char) 0);
            ridx = 0;
            widx = 0;
        } else {
            System.arraycopy(buffer, shiftCount, buffer, 0, buffer.length - shiftCount);
            ridx = Math.max(0, ridx - shiftCount);
            widx = Math.max(0, widx - shiftCount);
//            Arrays.fill(buffer, widx, buffer.length, (char) 0);
        }
    }

    public void grow(int capacity) {
        char[] buffer = this.buffer;
        if (capacity <= buffer.length) {
            return;
        }
        this.buffer = Arrays.copyOf(this.buffer, capacity);
    }

    @Override
    public int length() {
        return Math.max(0, widx - ridx);
    }

    @Override
    public char charAt(int index) {
        return buffer[ridx + index];
    }

    @Nonnull
    @Override
    public CharSequence subSequence(int start, int end) {
        return new String(buffer, ridx + start, end - start);
    }

    @Nonnull
    @Override
    public String toString() {
        return "CharBuffer{" +
                "buffer='" + encodeBuffer() + "'" +
                ", ridx=" + ridx +
                ", widx=" + widx +
                '}';
    }

    private String encodeBuffer() {
        if (ridx >= widx) {
            return "";
        }
        return new String(buffer, ridx, Math.max(0, widx - ridx));
    }
}

File: csharp/Dson/src/Text/CharBuffer.cs
#region LICENSE

//  Copyright 2023 wjybxx(845740757@qq.com)
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to iBn writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#endregion

using System;
using System.IO;
using Wjybxx.Commons.Collections;
using Wjybxx.Dson.IO;

namespace Wjybxx.Dson.Text;

internal class CharBuffer
{
    internal char[] buffer;
    internal int ridx;
    internal int widx;

    internal CharBuffer(int length) {
        this.buffer = new char[length];
    }

    internal CharBuffer(char[] buffer) {
        this.buffer = buffer ?? throw new ArgumentNullException(nameof(buffer));
    }

    public int Capacity => buffer.Length;

    public bool IsReadable => ridx < widx;

    public bool IsWritable => widx < buffer.Length;

    public int WritableChars => buffer.Length - widx;

    public int ReadableChars => Math.Max(0, widx - ridx);

    /** Length为可读字节数 */
    public int Length => Math.Max(0, widx - ridx);

    public char CharAt(int index) {
        return buffer[ridx + index];
    }

    #region 读写

    public char Read() {
        if (ridx == widx) throw new InternalBufferOverflowException();
        return buffer[ridx++];
    }

    public void Unread() {
        if (ridx == 0) throw new InternalBufferOverflowException();
        ridx--;
    }

    public void Write(char c) {
        if (widx == buffer.Length) {
            throw new InternalBufferOverflowException();
        }
        buffer[widx++] = c;
    }

    public void Write(char[] chars) {
        if (chars.Length == 0) {
            return;
        }
        if (widx + chars.Length > buffer.Length) {
            throw new InternalBufferOverflowException();
        }
        Array.Copy(chars, 0, buffer, widx, chars.Length);
        widx += chars.Length;
    }

    public void Write(char[] chars, int offset, int len) {
        if (len == 0) {
            return;
        }
        BinaryUtils.CheckBuffer(chars.Length, offset, len);
        if (widx + len > buffer.Length) {
            throw new InternalBufferOverflowException();
        }
        Array.Copy(chars, offset, buffer, widx, len);
        widx += len;
    }

    /**
     * 将给定buffer中的可读字符写入到当前buffer中
     * 给定的buffer的读索引将更新，当前buffer的写索引将更新
     *
     * @return 写入的字符数；返回0时可能是因为当前buffer已满，或给定的buffer无可读字符
     */
    public int Write(CharBuffer charBuffer) {
        int n = Math.Min(WritableChars, charBuffer.ReadableChars);
        if (n == 0) {
            return 0;
        }
        Write(charBuffer.buffer, charBuffer.ridx, n);
        charBuffer.AddRidx(n);
        return n;
    }

    #endregion

    #region 索引调整

    public void AddRidx(int count) {
        SetRidx(ridx + count);
    }

    public void AddWidx(int count) {
        SetWidx(widx + count);
    }

    public void SetRidx(int ridx) {
        if (ridx < 0 || ridx > widx) {
            throw new ArgumentException("ridx overflow");
        }
        this.ridx = ridx;
    }

    public void SetWidx(int widx) {
        if (widx < ridx || widx > buffer.Length) {
            throw new ArgumentException("widx overflow");
        }
        this.widx = widx;
    }

    public void SetIndexes(int ridx, int widx) {
        if (ridx < 0 || ridx > widx) {
            throw new ArgumentException("ridx overflow");
        }
        if (widx > buffer.Length) {
            throw new ArgumentException("widx overflow");
        }
        this.ridx = ridx;
        this.widx = widx;
    }

    #endregion

    #region 容量调整

    public void Shift(int shiftCount) {
        if (shiftCount <= 0) {
            return;
        }
        if (shiftCount >= buffer.Length) {
            ridx = 0;
            widx = 0;
        } else {
            Array.Copy(buffer, shiftCount, buffer, 0, buffer.Length - shiftCount);
            ridx = Math.Max(0, ridx - shiftCount);
            widx = Math.Max(0, widx - shiftCount);
        }
    }

    public void Grow(int capacity) {
        char[] buffer = this.buffer;
        if (capacity <= buffer.Length) {
            return;
        }
        this.buffer = CollectionUtil.CopyOf(this.buffer, 0, capacity);
    }

    #endregion

    public override string ToString() {
        return "CharBuffer{" +
               "buffer='" + EncodeBuffer() + "'" +
               ", ridx=" + ridx +
               ", widx=" + widx +
               '}';
    }

    private string EncodeBuffer() {
        if (ridx >= widx) {
            return "";
        }
        return new string(buffer, ridx, Math.Max(0, widx - ridx));
    }
}